### PR TITLE
Allow `EntryPointOffset` to be serde'd from int/string

### PR DESCRIPTION
**Pull request by giladchase**
- In new releases offsets will be numbers, whereas in the past they were
  strings.
- Simplified casting logic.
- Remove `serde from=` since it isn't relevant.
- Added a test for the `into` case.

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 5 unresolved discussions (waiting on @giladchase and @yair-starkware)

___
*[`src/deprecated_contract_class.rs` line 152 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-T9UC0BuLYJbm_UGu:-NR-T9UC0BuLYJbm_UGv:bq8ejzs) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L152)):*
> ```XML
> }
> 
> pub fn num_or_string<'de, D: Deserializer<'de>>(deserializer: D) -> Result<usize, D::Error> {
> ```

Either `str` or `string` everywhere?



_Suggestion:_
```XML
int_or_string
```
___
*[`src/deprecated_contract_class.rs` line 153 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-TGVM1yQaOlzcLo5p:-NR-TGVN1ecEOSBS7plY:undefined) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L153)):*
> ```XML
> 
> pub fn num_or_string<'de, D: Deserializer<'de>>(deserializer: D) -> Result<usize, D::Error> {
>     let usize_val = match Value::deserialize(deserializer)? {
> ```





_Suggestion:_
```XML
usize_value
```
___
*[`src/deprecated_contract_class.rs` line 154 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-TIfN1Fq0AUO6m3t7:-NR-TIfN1Fq0AUO6m3t8:undefined) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L154)):*
> ```XML
> pub fn num_or_string<'de, D: Deserializer<'de>>(deserializer: D) -> Result<usize, D::Error> {
>     let usize_val = match Value::deserialize(deserializer)? {
>         Value::Number(num) => {
> ```





_Suggestion:_
```XML
number
```
___
*[`src/state_test.rs` line 20 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-TjJW092rljY0lTd_:-NR-TjJW092rljY0lTda:btk7n2l) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/state_test.rs#L20)):*
> ```XML
>             "offset_2": "0x7b"
>         }"#;
>     type Offsets = HashMap<String, EntryPointOffset>;
> ```

Why is this type needed?



_Code quote:_
```XML
type Offsets = HashMap<String, EntryPointOffset>;
```
___
*[`src/state_test.rs` line 28 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-Td9C0C7a8DGw5KxF:-NR-Td9C0C7a8DGw5KxG:bwcryig) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/state_test.rs#L28)):*
> ```XML
> 
> #[test]
> fn entry_point_offset_into_json_str() {
> ```

I think we want to dump as `int`, no?



_Code quote:_
```XML
entry_point_offset_into_json_str
```

**Review by yair-starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 6 unresolved discussions (waiting on @giladchase)

___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NR2CxhI8-uzh1ez91ZG:b6cpiz9) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
> ```XML
> )]
> #[serde(into = "String")]
> pub struct EntryPointOffset(#[serde(deserialize_with = "num_or_string")] pub usize);
> ```

Why do you need a deserialization implementation instead of just modifying the TryFrom implementation?

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 4 unresolved discussions (waiting on @elintul and @yair-starkware)

___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NRSuyw2-gsZ2_JK4QKq:b2leea3) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
<details><summary><i>Previously, yair-starkware (Yair) wrote…</i></summary><blockquote>

Why do you need a deserialization implementation instead of just modifying the TryFrom implementation?
</blockquote></details>

Cause AFAIK we can do`#[serde(try_from = (String,Number),...)]`.

In other words, I need some way to tell serde to `try_from` both of these options, so `deserialize_with` seemed to fit.
Is there a better way?
___
*[`src/deprecated_contract_class.rs` line 152 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-T9UC0BuLYJbm_UGu:-NRSxbCD79V1je48wKmA:b-3ir8o5) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L152)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Either `str` or `string` everywhere?
</blockquote></details>

Sure about this one? 
In the function, we deserialize the input into `Number`, which is a serde type that is strictly more general than an int.
___
*[`src/state_test.rs` line 20 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-TjJW092rljY0lTd_:-NRSwn5nCnDMS4Cb1MwF:b-ql98re) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/state_test.rs#L20)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Why is this type needed?
</blockquote></details>

Wasn't really... just the type annotation was needed (for serde_json).
Fixed.
___
*[`src/state_test.rs` line 28 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-Td9C0C7a8DGw5KxF:-NRSwxlD7GbYhxz3NvnJ:buv5rg6) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/state_test.rs#L28)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I think we want to dump as `int`, no?
</blockquote></details>

I guess, but it's a breaking change for papyrus maybe? @yair-starkware WDYT?

**Review by giladchase — COMMENTED** [review]
Not sure why this is showing the revert commit, on github it looks OK...

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-:-NRT2PIS6bowA1S2jEp1:bt1ngbp)* status: 0 of 17 files reviewed, 4 unresolved discussions (waiting on @elintul and @yair-starkware)

File: src/state.rs [hunk 1]
```diff
@@ -125,6 +122,22 @@ impl From<EntryPointOffset> for String {
...
+        Value::Number(number) => {
+            number.as_u64().ok_or(DeserializationError::custom("Cannot cast number to usize."))?
+                as usize
+        }
+        Value::String(s) => hex_str_try_into_usize(&s).map_err(DeserializationError::custom)?,
+        _ => return Err(DeserializationError::custom("Cannot cast value into usize.")),
+    };
+    Ok(usize_value)
+}
+
```
Comment (yair-starkware) [inline src/state.rs:136]: Can you move the implementation of `hex_str_try_into_usize` into TryFrom<String> and use try_from inside num_or_string?

**Review by elintul — COMMENTED** [review]
> Reviewed 1 of 17 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: 1 of 17 files reviewed, 2 unresolved discussions (waiting on @yair-starkware)

File: src/state.rs [hunk 2]
```diff
@@ -125,6 +122,22 @@ impl From<EntryPointOffset> for String {
     }
 }
 
+pub fn num_or_string<'de, D: Deserializer<'de>>(deserializer: D) -> Result<usize, D::Error> {
```
Comment (elintul) [inline src/state.rs:125]: `num -> number`?
`string -> str`, or `string` everywhere?

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: 0 of 17 files reviewed, 4 unresolved discussions (waiting on @elintul and @yair-starkware)

___
*[`src/state.rs` line 125 at r6](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NRSy1IUteZesmoqU8mt-r6-125:-NRTF9Fh6Di5Dc7MwN8p:b-g326f6) ([raw file](https://github.com/starkware-libs/starknet-api/blob/a335ef7886a9809ec481990c0e1fd998a82d7a5e/src/state.rs#L125)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

`num -> number`?
`string -> str`, or `string` everywhere?
</blockquote></details>

Done for num -> number.

What did "string -> str, or string everywhere" mean?
___
*[`src/state.rs` line 136 at r6](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NRSy1IUteZesmoqU8mt-r6-136:-NRTFRL87x7INKZwhsy_:bwesb38) ([raw file](https://github.com/starkware-libs/starknet-api/blob/a335ef7886a9809ec481990c0e1fd998a82d7a5e/src/state.rs#L136)):*
<details><summary><i>Previously, yair-starkware (Yair) wrote…</i></summary><blockquote>

Can you move the implementation of `hex_str_try_into_usize` into TryFrom<String> and use try_from inside num_or_string?
</blockquote></details>

Not quite, the tryFrom outputs `EntryPointOffset` result objects, while the deserializer function outputs `usize`ones.
___
*[`src/state_test.rs` line 28 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-Td9C0C7a8DGw5KxF:-NRTEURB0l8c_uGJiZLQ:ba3vkz3) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/state_test.rs#L28)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

I guess, but it's a breaking change for papyrus maybe? @yair-starkware WDYT?
</blockquote></details>

Talked to Yair, this is probably not a problem so I changed into it (which mearnt deleting a bunch of code :)).
I'll make sure this doesn't break papyrus before merging.

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r6, 1 of 15 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: 3 of 17 files reviewed, 4 unresolved discussions (waiting on @giladchase and @yair-starkware)

___
*[`src/lib.rs` line 23 at r8](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NRTJupf1gA_8ab8Ym0c:-NRTJupf1gA_8ab8Ym0d:b-suv37h) ([raw file](https://github.com/starkware-libs/starknet-api/blob/49d3adb3d30f35c370e6a4ef52169d8804857695/src/lib.rs#L23)):*
> ```XML
>     InnerDeserialization(#[from] InnerDeserializationError),
>     /// Error when serializing into number.
>     #[error(transparent)]
> ```

Alphabetize?



_Code quote:_
```XML
    /// Error when serializing into number.
    #[error(transparent)]
    ParseIntError(#[from] ParseIntError),
    #[error("Out of range {string}.")]
    /// An error for when a value is out of range.
    OutOfRange { string: String },
```
___
*[`src/state.rs` line 125 at r6](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NRSy1IUteZesmoqU8mt-r6-125:-NRTK-FHDq8s_JDcVeO5:b-gcmcnk) ([raw file](https://github.com/starkware-libs/starknet-api/blob/a335ef7886a9809ec481990c0e1fd998a82d7a5e/src/state.rs#L125)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Done for num -> number.

What did "string -> str, or string everywhere" mean?
</blockquote></details>

`hex_string`

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: 1 of 17 files reviewed, 3 unresolved discussions (waiting on @elintul and @yair-starkware)

___
*[`src/lib.rs` line 23 at r8](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NRTJupf1gA_8ab8Ym0c:-NRTKJRm2q1CFgvX_YtF:bb6un1j) ([raw file](https://github.com/starkware-libs/starknet-api/blob/49d3adb3d30f35c370e6a4ef52169d8804857695/src/lib.rs#L23)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Alphabetize?
</blockquote></details>

Done
___
*[`src/state.rs` line 125 at r6](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NRSy1IUteZesmoqU8mt-r6-125:-NRTKYNZ0RunHNdgKk_X:b-1hiun6) ([raw file](https://github.com/starkware-libs/starknet-api/blob/a335ef7886a9809ec481990c0e1fd998a82d7a5e/src/state.rs#L125)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

`hex_string`
</blockquote></details>

Oh i see, Done.

**Review by elintul — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-:-NRTLByt2ZYTrW7haSuw:bnfp4nl)* status: 3 of 17 files reviewed, 2 unresolved discussions (waiting on @yair-starkware)

**Review by elintul — APPROVED** [review]
> Reviewed 1 of 4 files at r6, 13 of 15 files at r7.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 2 unresolved discussions (waiting on @yair-starkware)

**Review by yair-starkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 15 files at r7, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-:-NRTMtOP0ED_AcQvvCI5:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @giladchase)

___
*[`src/state_test.rs` line 28 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR-Td9C0C7a8DGw5KxF:-NRT1asq1IE3ypunNCkq:b-42qlmq) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/state_test.rs#L28)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

I guess, but it's a breaking change for papyrus maybe? @yair-starkware WDYT?
</blockquote></details>

I don't think it will break us because we use this struct as-is in the client,
___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NRSzmqMA7rNCBNhKqpA:b-hopi9n) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Cause AFAIK we can do`#[serde(try_from = (String,Number),...)]`.

In other words, I need some way to tell serde to `try_from` both of these options, so `deserialize_with` seemed to fit.
Is there a better way?
</blockquote></details>

Why do you need both options and not just Number? Is it only for new blocks?

**Review by elintul — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 1 unresolved discussion (waiting on @giladchase)

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 1 unresolved discussion (waiting on @elintul)

___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NRTT8xl-bpIKov7t2vH:b-fhd6wy) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
<details><summary><i>Previously, yair-starkware (Yair) wrote…</i></summary><blockquote>

Why do you need both options and not just Number? Is it only for new blocks?
</blockquote></details>

Backwards compatibility for us from python (that right @elintul ?)

**Review by elintul — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 1 unresolved discussion (waiting on @yair-starkware)

___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NRTTJSt2OJdF_YnQMHr:b2wh4q8) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Backwards compatibility for us from python (that right @elintul ?) 
</blockquote></details>

Yes, we read old raw compiled contracts from Aerospike.

**Review by elintul — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 1 unresolved discussion (waiting on @yair-starkware)

___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NRTTUtmE6TrJZmMHX9-:b-hipcu2) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Yes, we read old raw compiled contracts from Aerospike.
</blockquote></details>

Let's move this to Slack, I want to merge this PR since it's a bugfix I'm waiting for.

**Review by giladchase — COMMENTED** [review]
> Dismissed @yair-starkware from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: all files reviewed, 1 unresolved discussion (waiting on @elintul)

___
*[`src/deprecated_contract_class.rs` line 137 at r2](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-NR2CxhI8-uzh1ez91ZF:-NRTTeoZ06tV2BVJ0oow:betove5) ([raw file](https://github.com/starkware-libs/starknet-api/blob/96a788879e9f9a90f3b8d9ab1e66ce69dba040ad/src/deprecated_contract_class.rs#L137)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Let's move this to Slack, I want to merge this PR since it's a bugfix I'm waiting for.
</blockquote></details>

OK, dismissing block @yair-starkware  since we talked about this offline already and you're AFK.

**Review by elintul — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @giladchase)

**Review by elintul — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-api/33#-:-NRTTyNF2-BQAMVb6pX-:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @giladchase)
